br/pkg/lightning/backend/local/local.go

```diff
@@ -1294,6 +1300,123 @@ func (local *local) readAndSplitIntoRange(ctx context.Context, engine *Engine, r
 	return ranges, nil
 }
 
+func (local *local) scanRegion(
+	ctxt context.Context,
+	engine *Engine,
+	start, end []byte,
+	regionSplitSize int64,
+	regionSplitKeys int64,
+) ([]*split.RegionInfo, error) {
+	ito := &pebble.IterOptions{
+		LowerBound: start,
+		UpperBound: end,
+	}
+
+	iter := engine.newKVIter(ctxt, ito)
+	//nolint: errcheck
+	defer iter.Close()
+	// Needs seek to first because NewIter returns an iterator that is unpositioned
+	hasKey := iter.First()
+	if iter.Error() != nil {
+		return nil, errors.Annotate(iter.Error(), "failed to read the first key")
+	}
+	if !hasKey {
+		log.FromContext(ctxt).Info("There is no pairs in iterator",
+			logutil.Key("start", start),
+			logutil.Key("end", end))
+		engine.finishedRanges.add(Range{start: start, end: end})
+		return nil, nil
+	}
+	pairStart := append([]byte{}, iter.Key()...)
+	iter.Last()
+	if iter.Error() != nil {
+		return nil, errors.Annotate(iter.Error(), "failed to seek to the last key")
+	}
+	pairEnd := append([]byte{}, iter.Key()...)
+
+	var regions []*split.RegionInfo
+	var err error
+	ctx, cancel := context.WithCancel(ctxt)
+	defer cancel()
+
+WriteAndIngest:
+	for retry := 0; retry < maxRetryTimes; {
+		if retry != 0 {
+			select {
+			case <-time.After(time.Second):
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
+		}
+		startKey := codec.EncodeBytes([]byte{}, pairStart)
+		endKey := codec.EncodeBytes([]byte{}, nextKey(pairEnd))
+		regions, err = split.PaginateScanRegion(ctx, local.splitCli, startKey, endKey, scanRegionLimit)
+		if err != nil || len(regions) == 0 {
+			log.FromContext(ctx).Warn("scan region failed", log.ShortError(err), zap.Int("region_len", len(regions)),
+				logutil.Key("startKey", startKey), logutil.Key("endKey", endKey), zap.Int("retry", retry))
+			retry++
+			continue WriteAndIngest
+		}
+		return regions, nil
+	}
+
+	return nil, err
+}
+
+func pickRanges(rangeAndRegions []RangeAndRegion, regionProcessingMap *sync.Map, finished int) *RangeAndRegion {
+	minScore := math.MaxInt64
+	var bestRange *RangeAndRegion
+	index := 0
+	for i, rangeAndRegion := range rangeAndRegions {
+		if !rangeAndRegion.isScheduled {
```

Comment:
to avoid iterating scheduled `rangeAndRegion` over and over again, how about we use map type for `rangeAndRegions` to efficiently delete the schedued item?